src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -234,7 +234,7 @@ private static uint HandleEventConnected(State state, ref ConnectionEvent connec
 
         private static uint HandleEventShutdownInitiatedByTransport(State state, ref ConnectionEvent connectionEvent)
         {
-            if (!state.Connected && state.ConnectTcs != null)
+            if (!state.Connected && state.ConnectTcs != null && !state.ConnectTcs.Task.IsCompleted)
```

Comment:
Your PR description suggests there's a race condition?  If that's the case, then couldn't the SetException a few lines down fail if the task is completed between here and there?

Reply:
The race I see is setting the result (and previously setting ConnectTcs to null) and returning the task in ConnectAsync. MsQuic will queue all the events so there should not be race between processing two events.

Reply:
Ok